infrastructure/src/main/java/org/corfudb/infrastructure/management/HealingDetector.java

```diff
@@ -0,0 +1,235 @@
+            } catch (NetworkException ne) {
+                log.debug("Error creating router for {}", members[i]);
+            }
+        }
+        pollCompletableFutures = new CompletableFuture[members.length];
+    }
+
+    /**
+     * Each PollRound executes "detectionThreshold" number of iterations and generates the
+     * poll report.
+     *
+     * @return Poll Report with detected healed nodes.
+     */
+    private PollReport pollRound() {
+
+        Set<String> membersSet = new HashSet<>(Arrays.asList(members));
+        Set<String> healed = new HashSet<>();
+        resetResponses();
+
+        // In each iteration we poll all the servers in the members list.
+        for (int iteration = 0; iteration < detectionThreshold; iteration++) {
+
+            // Ping all nodes and await their responses.
+            pollOnceAsync();
+
+            // Collect all poll responses.
+            collectResponsesAndVerifyEpochs(iteration);
+
+            if (!isHealedNodesPresent(membersSet)) {
+                break;
+            }
+
+            Sleep.MILLISECONDS.sleepUninterruptibly(interIterationInterval);
+        }
+
+        // Check all responses and  report all healed nodes.
+        for (int i = 0; i < members.length; i++) {
+            if (responses[i] == (detectionThreshold - 1)) {
+                healed.add(members[i]);
+            }
+        }
+
+        return new PollReport.PollReportBuilder()
+                .pollEpoch(currentEpoch)
+                .healingNodes(healed)
+                .build();
+    }
+
+    /**
+     * Reset all responses to an invalid iteration number, -1.
+     */
+    private void resetResponses() {
+        for (int i = 0; i < responses.length; i++) {
+            responses[i] = INVALID_RESPONSE;
+        }
+    }
+
+    /**
+     * Poll all members servers once asynchronously and store their futures in
+     * pollCompletableFutures.
+     */
+    private void pollOnceAsync() {
+
+        // Poll servers for health.  All ping activity will happen in the background.
+        for (int i = 0; i < members.length; i++) {
+            try {
+                pollCompletableFutures[i] = membersRouters[i].getClient(BaseClient.class).ping();
+            } catch (Exception e) {
+                CompletableFuture<Boolean> cf = new CompletableFuture<>();
+                cf.completeExceptionally(e);
+                pollCompletableFutures[i] = cf;
+            }
+        }
+    }
+
+    /**
+     * Block on all poll futures and collect the responses. There are 3 possible cases.
+     * 1. Receive a PONG. We set the responses[i] to the polling iteration number.
+     * 2. WrongEpochException is thrown. We mark as a successful response.
+     * 3. Other Exception is thrown. We do nothing. (The response[i] in this case is left behind.)
+     *
+     * @param pollIteration poll iteration in the ongoing polling round.
+     */
+    private void collectResponsesAndVerifyEpochs(int pollIteration) {
```

Comment:
Well, it seems this function is unnecessary too, because you can just do the transformation from the previous comment in one step, adding a helper function for clarity:

```java
boolean pingSuccessful(@Nonnull CompletableFuture<Boolean> future) {
     try {
        return future.get();
      } catch (Exception e) {
        if (e.getCause() instanceof WrongEpochException) {
             return true;
        }
        return false;
     }
}
```

```java
        // Iterate through the array
        IntStream.range(0, routers.length)
            .boxed()
            // Convert to a map of routers -> pings
            .collect(Collectors.toMap(i -> i,
                i -> pingSuccessful(routers[0].getClient(BaseClient.class).ping())))
            // If the ping was successful, update the iteration
            .forEach(e -> {
                responses[e.getKey()] = e.getValue() ? pollIteration : responses[e.getKey()];
            });
```